File: src/PwrDrvr.MicroApps.Deployer/Controllers/VersionController.cs
using PwrDrvr.MicroApps.DataLib;
using PwrDrvr.MicroApps.Deployer.Lambda;
using Microsoft.AspNetCore.Mvc;
using System.Threading.Tasks;
using Amazon.S3;
using Amazon.Lambda;
using Amazon.Lambda.Model;
using System;
using System.Text.Json;
using Amazon.ApiGatewayV2;
using Amazon.ApiGatewayV2.Model;

namespace PwrDrvr.MicroApps.Deployer.Controllers {
  public class VersionBody {
    public string appName { get; set; }
    public string semVer { get; set; }
    public string s3SourceURI { get; set; }
    public string lambdaARN { get; set; }
    public string defaultFile { get; set; }
  }

  [ApiController]
  [Route("deployer/[controller]")]
  public class VersionController : ControllerBase {
    // GET /deployer/version/{appName}/{semVer}
    [HttpGet("{appName}/{version}")]
    async public Task Get(string appName, string version) {
      try {
        // Check if the version exists
        var record = await Manager.GetAppVersion(appName, version);
        if (record != null && record.Status != "pending") {
          Response.StatusCode = 200;
          Console.WriteLine("App/Version already exists: {0}/{1}", appName, version);
          return;
        } else {
          Response.StatusCode = 404;
          Console.WriteLine("App/Version does not exist: {0}/{1}", appName, version);
          return;
        }
      } catch (Exception ex) {
        Response.StatusCode = 500;
        Console.WriteLine("Caught unexpected exception: {0}", ex.Message);
      }
    }

    // POST /deployer/version/
    [HttpPost]
    async public Task Post([FromBody] VersionBody versionBody) {
      try {
        Console.WriteLine("Got Body: {0}", JsonSerializer.Serialize(versionBody));

        const string destinationBucket = "pwrdrvr-apps";
        string destinationPrefix = string.Format("{0}/{1}", versionBody.appName, versionBody.semVer);

        // Check if the version exists
        var record = await Manager.GetAppVersion(versionBody.appName, versionBody.semVer);
        if (record != null && record.Status == "routed") {
          Response.StatusCode = 409;
          Console.WriteLine("App/Version already exists: {0}/{1}", versionBody.appName, versionBody.semVer);
          return;
        }

        // Create the version record
        if (record == null) {
          record = new DataLib.Models.Version() {
            AppName = versionBody.appName,
            SemVer = versionBody.semVer,
            Type = "lambda",
            Status = "pending",
            DefaultFile = versionBody.defaultFile,
          };
          // Save record with pending status
          await Manager.CreateVersion(record);
        }

        // Only copy the files if not copied yet
        if (record.Status == "pending") {
          var s3Client = new AmazonS3Client();

          // Parse the S3 Source URI
          var uri = new Uri(versionBody.s3SourceURI);

          var sourceBucket = uri.Host;
          var sourcePrefix = uri.AbsolutePath.Length >= 1 ? uri.AbsolutePath.Substring(1) : null;

          // Example Source: s3://pwrdrvr-apps-staging/release/1.0.0/
          var paginator = s3Client.Paginators.ListObjectsV2(new Amazon.S3.Model.ListObjectsV2Request() {
            BucketName = uri.Host,
            Prefix = sourcePrefix,
          });

          // Loop through all S3 source assets and copy to the destination
          await foreach (var obj in paginator.S3Objects) {
            var sourceKeyRootless = obj.Key.Remove(0, sourcePrefix.Length);
            // Console.WriteLine("object: ${0}", obj.Key);
            await s3Client.CopyObjectAsync(sourceBucket, obj.Key,
            destinationBucket, string.Format("{0}/{1}", destinationPrefix, sourceKeyRootless));
          }

          // Update status to assets-copied
          record.Status = "assets-copied";
          await Manager.CreateVersion(record);
        }

        // TODO: Confirm the Lambda Function exists
        var lambdaClient = new AmazonLambdaClient();
        // lambdaClient.CreateAliasAsync(new CreateAliasRequest() {
        //   FunctionName = "",
        //   FunctionVersion = "1",
        //   Name = string.Format("v{0}", versionBody.semVer.Replace('.', '_')),
        // });
        // var aliasResponse = await lambdaClient.GetAliasAsync(new GetAliasRequest() {
        //   FunctionName = "",
        //   Name = "",
        // });

        // Get the API Gateway
        var apigwy = new AmazonApiGatewayV2Client();
        var api = await GatewayInfo.GetAPI(apigwy);

        if (record.Status == "assets-copied") {
          // Get the account ID
          var lambdaArnParts = versionBody.lambdaARN.Split(':');
          var accountId = lambdaArnParts[4];
          var region = lambdaArnParts[3];

          // Ensure that the Lambda function allows API Gateway to invoke
          await lambdaClient.RemovePermissionAsync(new RemovePermissionRequest() {
            FunctionName = versionBody.lambdaARN,
            StatementId = "apigwy",
          });
          await lambdaClient.AddPermissionAsync(new AddPermissionRequest() {
            Principal = "apigateway.amazonaws.com",
            StatementId = "apigwy",
            Action = "lambda:InvokeFunction",
            FunctionName = versionBody.lambdaARN,
            SourceArn = string.Format("arn:aws:execute-api:{0}:{1}:{2}/*/*/{3}/*/api/{{proxy+}}", region, accountId, api.ApiId, versionBody.appName)
          });
          record.Status = "permissioned";
          await Manager.CreateVersion(record);
        }

        // Add Integration pointing to Lambda Function Alias
        var integrationId = "";
        if (record.Status == "permissioned") {
          if (!string.IsNullOrEmpty(record.IntegrationID)) {
            integrationId = record.IntegrationID;
          } else {
            var integration = await apigwy.CreateIntegrationAsync(new CreateIntegrationRequest() {
              ApiId = api.ApiId,
              IntegrationType = IntegrationType.AWS_PROXY,
              IntegrationMethod = "POST",
              PayloadFormatVersion = "2.0",
              IntegrationUri = versionBody.lambdaARN,
            });

            integrationId = integration.IntegrationId;

            // Save the created IntegrationID
            record.IntegrationID = integration.IntegrationId;
            record.Status = "integrated";
            await Manager.CreateVersion(record);
          }
        }

        // Add the route to API Gateway for appName/version/{proxy+}
        var routeRouter = await apigwy.CreateRouteAsync(new CreateRouteRequest() {
          ApiId = api.ApiId,
          Target = string.Format("integrations/{0}", integrationId),
          RouteKey = string.Format("ANY /{0}/{1}/api/{{proxy+}}", versionBody.appName, versionBody.semVer),
        });

        // Update the status - Final status
        record.Status = "routed";
        await Manager.CreateVersion(record);
      } catch (Exception ex) {
        Response.StatusCode = 500;
        Console.WriteLine("Caught unexpected exception: {0}", ex.Message);
      }
    }
  }
}

File: src/microapps-deployer/controllers/VersionController.ts
import {
  manager,
  IDeployVersionRequest,
  ICheckVersionExistsRequest,
  IDeployerResponse,
} from '../index';
import * as lambda from '@aws-sdk/client-lambda';
import * as s3 from '@aws-sdk/client-s3';
import * as apigwy from '@aws-sdk/client-apigatewayv2';
import GatewayInfo from '../lib/GatewayInfo';
import { Rules, Version } from '@pwrdrvr/microapps-datalib';
import Log from '../lib/Log';

const lambdaClient = new lambda.LambdaClient({});
const s3Client = new s3.S3Client({});
const apigwyClient = new apigwy.ApiGatewayV2Client({});

export default class VersionController {
  static destinationBucket = 'pwrdrvr-apps';

  public static async CheckVersionExists({
    appName,
    semVer,
  }: ICheckVersionExistsRequest): Promise<IDeployerResponse> {
    // Check if the version exists
    const record = await Version.LoadVersionAsync(manager.DBDocClient, appName, semVer);
    if (record !== undefined && record.Status !== 'pending') {
      Log.Instance.info('App/Version already exists', { appName, semVer });
      return { statusCode: 200 };
    } else {
      Log.Instance.info('App/Version does not exist', { appName, semVer });
      return { statusCode: 404 };
    }
  }

  public static async DeployVersion(request: IDeployVersionRequest): Promise<IDeployerResponse> {
    Log.Instance.debug(`Got Body:`, request);

    const destinationPrefix = `${request.appName}/${request.semVer}`.toLowerCase();

    // Check if the version exists
    let record = await Version.LoadVersionAsync(
      manager.DBDocClient,
      request.appName,
      request.semVer,
    );
    if (record !== undefined && record.Status === 'routed') {
      Log.Instance.info('App/Version already exists', {
        appName: request.appName,
        semVer: request.semVer,
      });
      return { statusCode: 409 };
    }

    // Create the version record
    if (record === undefined) {
      record = new Version({
        AppName: request.appName,
        SemVer: request.semVer,
        IntegrationID: '',
        Type: 'lambda',
        Status: 'pending',
        DefaultFile: request.defaultFile,
      });

      // Save record with pending status
      await record.SaveAsync(manager.DBDocClient);
    }

    // Only copy the files if not copied yet
    if (record.Status === 'pending') {
      // Parse the S3 Source URI
      const uri = new URL(request.s3SourceURI);

      const sourceBucket = uri.host;
      const sourcePrefix = uri.pathname.length >= 1 ? uri.pathname.slice(1) : '';

      // Example Source: s3://pwrdrvr-apps-staging/release/1.0.0/
      // Loop through all S3 source assets and copy to the destination
      await VersionController.CopyToProdBucket(sourceBucket, sourcePrefix, destinationPrefix);

      // Update status to assets-copied
      record.Status = 'assets-copied';
      await record.SaveAsync(manager.DBDocClient);
    }

    // TODO: Confirm the Lambda Function exists

    // Get the API Gateway
    const api = await GatewayInfo.GetAPI(apigwyClient);

    if (record.Status === 'assets-copied') {
      // Get the account ID
      const lambdaArnParts = request.lambdaARN.split(':');
      const accountId = lambdaArnParts[4];
      const region = lambdaArnParts[3];

      // Ensure that the Lambda function allows API Gateway to invoke
      try {
        await lambdaClient.send(
          new lambda.RemovePermissionCommand({
            FunctionName: request.lambdaARN,
            StatementId: 'microapps-version-root',
          }),
        );
      } catch {
        // Don't care if this remove throws
      }
      try {
        await lambdaClient.send(
          new lambda.RemovePermissionCommand({
            FunctionName: request.lambdaARN,
            StatementId: 'microapps-version',
          }),
        );
      } catch {
        // Don't care if this remove throws
      }
      await lambdaClient.send(
        new lambda.AddPermissionCommand({
          Principal: 'apigateway.amazonaws.com',
          StatementId: 'microapps-version-root',
          Action: 'lambda:InvokeFunction',
          FunctionName: request.lambdaARN,
          SourceArn: `arn:aws:execute-api:${region}:${accountId}:${api.ApiId}/*/*/${request.appName}/${request.semVer}`,
        }),
      );
      await lambdaClient.send(
        new lambda.AddPermissionCommand({
          Principal: 'apigateway.amazonaws.com',
          StatementId: 'microapps-version',
          Action: 'lambda:InvokeFunction',
          FunctionName: request.lambdaARN,
          SourceArn: `arn:aws:execute-api:${region}:${accountId}:${api.ApiId}/*/*/${request.appName}/${request.semVer}/{{proxy+}}`,
        }),
      );
      record.Status = 'permissioned';
      await record.SaveAsync(manager.DBDocClient);
    }

    // Add Integration pointing to Lambda Function Alias
    let integrationId = '';
    if (record.Status === 'permissioned') {
      if (record.IntegrationID !== undefined && record.IntegrationID !== '') {
        integrationId = record.IntegrationID;
      } else {
        const integration = await apigwyClient.send(
          new apigwy.CreateIntegrationCommand({
            ApiId: api.ApiId,
            IntegrationType: apigwy.IntegrationType.AWS_PROXY,
            IntegrationMethod: 'POST',
            PayloadFormatVersion: '2.0',
            IntegrationUri: request.lambdaARN,
          }),
        );

        integrationId = integration.IntegrationId as string;

        // Save the created IntegrationID
        record.IntegrationID = integration.IntegrationId as string;
        record.Status = 'integrated';
        await record.SaveAsync(manager.DBDocClient);
      }
    }

    if (record.Status === 'integrated') {
      // Add the routes to API Gateway for appName/version/{proxy+}
      try {
        await apigwyClient.send(
          new apigwy.CreateRouteCommand({
            ApiId: api.ApiId,
            Target: `integrations/${integrationId}`,
            RouteKey: `ANY /${request.appName}/${request.semVer}`,
          }),
        );
      } catch (err) {
        // Don't care
        Log.Instance.error('Caught unexpected error on app/ver route add');
        Log.Instance.error(err);
      }
      try {
        await apigwyClient.send(
          new apigwy.CreateRouteCommand({
            ApiId: api.ApiId,
            Target: `integrations/${integrationId}`,
            RouteKey: `ANY /${request.appName}/${request.semVer}/{{proxy+}}`,
          }),
        );
      } catch (err) {
        // Don't care
        Log.Instance.error('Caught unexpected error on {{proxy+}} route add');
        Log.Instance.error(err);
      }

      // Update the status - Final status
      record.Status = 'routed';
      await record.SaveAsync(manager.DBDocClient);
    }

    // Check if there are any release rules
    // If no rules record, create one pointing to this version by default
    let rules = await Rules.LoadAsync(manager.DBDocClient, request.appName);
    if (rules === undefined) {
      rules = new Rules({
        AppName: request.appName,
        RuleSet: {},
        Version: 1,
      });
      rules.RuleSet.default = {
        SemVer: request.semVer,
        AttributeName: '',
        AttributeValue: '',
      };
      await rules.SaveAsync(manager.DBDocClient);
    }

    return { statusCode: 201 };
  }

  private static async CopyFilesInList(
    list: s3.ListObjectsV2CommandOutput,
    sourceBucket: string,
    sourcePrefix: string,
    destinationPrefix: string,
  ): Promise<void> {
    if (list === undefined || list.Contents === undefined) {
      return;
    }
    for (const obj of list.Contents) {
      const sourceKeyRootless = obj.Key?.slice(sourcePrefix.length);
      await s3Client.send(
        new s3.CopyObjectCommand({
          // Source
          CopySource: `${sourceBucket}/${obj.Key}`,
          // Destination
          Bucket: VersionController.destinationBucket,
          Key: `${destinationPrefix}/${sourceKeyRootless}`,
        }),
      );
    }
  }

  private static async CopyToProdBucket(
    sourceBucket: string,
    sourcePrefix: string,
    destinationPrefix: string,
  ) {
    let list = await s3Client.send(
      new s3.ListObjectsV2Command({
        Bucket: sourceBucket,
        Prefix: sourcePrefix,
      }),
    );
    await VersionController.CopyFilesInList(list, sourceBucket, sourcePrefix, destinationPrefix);
    while (list.IsTruncated) {
      list = await s3Client.send(
        new s3.ListObjectsV2Command({
          Bucket: sourceBucket,
          Prefix: sourcePrefix,
          ContinuationToken: list.NextContinuationToken,
        }),
      );
      await VersionController.CopyFilesInList(list, sourceBucket, sourcePrefix, destinationPrefix);
    }
  }
}
